File: Example 21 JNA/src/example21.c
// Projeto : JNA Example #21
// Algoritmo: example21.c
// Função : : Receive an array of unions from C
// Autor(es) : Hugo Dionizio Santos (adaptações e correções), Ethan Shayne in http://www.eshayne.com/jnaex
// Data : Sáb Mai 16 16:27:45 BRT 2015

#include <stdlib.h>
#include <string.h>
#include <assert.h>

// Seção de Variáveis Globais

// Interface do algoritmo

// Função Principal
//	Seção de Variáveis Locais
//	Seção de Comandos

// Seção de Funções
typedef union Example21Union {
	int intnumber;
	float floatnumber;
	char* stringval;
} Example21Union;

enum Example21UnionType{
    EX21_UNKNOWN = 0,
    EX21_INTEGER = 1,
    EX21_FLOAT = 2,
    EX21_STRING = 3
};

typedef struct Example21UnionHolder {
	enum Example21UnionType uniontype;
	Example21Union unionval;
} Example21UnionHolder;

typedef struct Example21UnionList {
	int numVals;
	Example21UnionHolder* vals;
} Example21UnionList;

Example21UnionList example21_get()
{
	Example21UnionList list;
	list.numVals = 3;
	list.vals = (Example21UnionHolder*)malloc(sizeof(Example21UnionHolder) * 3);
	memset(list.vals, 0, sizeof(Example21UnionHolder) * 3);

	list.vals[0].uniontype = EX21_INTEGER;
	list.vals[0].unionval.intnumber = 33;
	list.vals[1].uniontype = EX21_FLOAT;
	list.vals[1].unionval.floatnumber = 34.5f;
	list.vals[2].uniontype = EX21_STRING;
	list.vals[2].unionval.stringval = (char*)malloc(sizeof(char) * 6);
	strcpy(list.vals[2].unionval.stringval, "hello");

	return list;
}

void example21_cleanup(Example21UnionList list)
{
	int i=0;

	if (0 == list.numVals)
	{
		assert(NULL == list.vals);
		return;
	}
	assert(NULL != list.vals);

	for (i=0; i<list.numVals; i++)
	{
		if (EX21_STRING == list.vals[i].uniontype)
		{
			assert(NULL != list.vals[i].unionval.stringval);
			free(list.vals[i].unionval.stringval);
		}
	}
	free(list.vals);
}

// Fim do algoritmo

File: Example 21 JNA/src/Example21.java
// Projeto : JNA Example #21
// Classe: Example21.java
// Função : : Receive an array of unions from C
// Autor(es) : Hugo Dionizio Santos
// Data : Sáb Mai 16 16:27:45 BRT 2015

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.ptr.PointerByReference;		
import com.sun.jna.Pointer;
import com.sun.jna.Structure;
import com.sun.jna.Structure.ByReference;
import com.sun.jna.Union;

public class Example21 {
	public interface CLibrary extends Library {
		public static interface Example21UnionType {
		    public static final int UNKNOWN = 0;
		    public static final int INTEGER = 1;
		    public static final int FLOAT = 2;
		    public static final int STRING = 3;
		}    
		public static class Example21Union extends Union {
			public static class ByValue extends Example21Union implements Union.ByValue {};
		
			public int intnumber;
			public float floatnumber;
			public String stringval;
		}
	
		public static class Example21UnionHolder extends Structure {
			public static class ByReference extends Example21UnionHolder implements Structure.ByReference {}
		
			public int uniontype;
			public Example21Union unionval;
		}
	
		public static class Example21UnionList extends Structure {
			public static class ByValue extends Example21UnionList implements Structure.ByValue {}
		
			public int numVals;
			public Example21UnionHolder.ByReference vals;
		}

		public Example21UnionList.ByValue example21_get();
		public void example21_cleanup(Example21UnionList.ByValue list);
	}

	static public void main(String argv[]) {
		CLibrary clib;
		CLibrary.Example21UnionList.ByValue ex21list;

		try {
			clib = (CLibrary)Native.loadLibrary("./libexample21.so", CLibrary.class);
			// 
			ex21list = clib.example21_get();
		}
		catch (UnsatisfiedLinkError e) {
			clib = (CLibrary)Native.loadLibrary("./bin/libexample21.so", CLibrary.class);
			// 
			ex21list = clib.example21_get();
		}

		ex21list.read();
		System.out.println("Example 21: received " + ex21list.numVals + " values:");
		CLibrary.Example21UnionHolder[] ex21holders = (CLibrary.Example21UnionHolder[])ex21list.vals.toArray(ex21list.numVals);
		for (int i=0; i<ex21list.numVals; i++)
		{
			switch (ex21holders[i].uniontype)
			{
				case CLibrary.Example21UnionType.INTEGER:
				{
					ex21holders[i].unionval.setType(Integer.TYPE);
					ex21holders[i].read();
					System.out.println("example 21: received integer value " + ex21holders[i].unionval.intnumber);
					break;
				}
				case CLibrary.Example21UnionType.FLOAT:
				{
					ex21holders[i].unionval.setType(Float.TYPE);
					ex21holders[i].read();
					System.out.println("example 21: received float value " + ex21holders[i].unionval.floatnumber);
					break;
				}
				case CLibrary.Example21UnionType.STRING:
				{
					ex21holders[i].unionval.setType(String.class);
					ex21holders[i].read();
					System.out.println("example 21: received string value " + ex21holders[i].unionval.stringval);
					break;
				}
				default:
				{
					System.err.println("example 21: error: unrecognized union type received: " + ex21holders[i].uniontype);
					break;
				}
			}
		}

		// cleanup
		System.out.println("Example 21 cleanup");
		ex21list.write();
		clib.example21_cleanup(ex21list);
	}
}

// Fim da classe
